File: java/src/quantity/Quantity.java
/*
 * Copyright (c) 2019 by Fred George
 * May be used freely except for training; license required for training.
 */

package quantity;

// Understands a specific measurement
public class Quantity {

    private final double amount;
    private final Unit unit;

    Quantity(double amount, Unit unit) {
        this.amount = amount;
        this.unit = unit;
    }

    @Override
    public boolean equals(Object other) {
        return this == other || other instanceof Quantity && this.equals((Quantity)other);
    }

    private boolean equals(Quantity other) {
        return this.isCompatible(other) && this.amount == convertedAmount(other);
    }

    private boolean isCompatible(Quantity other) {
        return this.unit.isCompatible(other.unit);
    }

    private double convertedAmount(Quantity other) {
        return this.unit.convertedAmount(other.amount, other.unit);
    }

    @Override
    public int hashCode() {
        return unit.hashCode(amount);
    }

    public Quantity negate() {
        return new Quantity(-amount, unit);
    }

    public Quantity plus(Quantity other) {
        return new Quantity(this.amount + convertedAmount(other), this.unit);
    }

    public Quantity minus(Quantity other) {
        return this.plus(other.negate());
    }
}

File: ruby/lib/quantity/quantity.rb
# Copyright (c) 2019 by Fred George
# May be used freely except for training; license required for training.
# @author Fred George  fredgeorge@acm.org

require_relative './unit'

# Understands a specific measurement
class Quantity
  attr_reader :amount, :unit
  protected :amount, :unit

  def initialize amount, unit
    @amount, @unit = amount.to_f, unit
  end

  def == other
    self.eql? other
  end

  def eql? other
    return false unless other.is_a? Quantity
    compatible?(other) && self.amount == converted_amount(other)
  end

  def hash
    @unit.hash_code @amount
  end

  def +@
    self
  end

  def -@
    Quantity.new -amount, unit
  end

  def + other
    Quantity.new self.amount + converted_amount(other), self.unit
  end

  def - other
    self + -other
  end

  private

    def compatible? other
      self.unit.compatible? other.unit
    end

    def converted_amount other
      self.unit.converted_amount other.amount, other.unit
    end

end
